src/mgr/BaseMgrModule.cc

```diff
@@ -107,7 +107,9 @@ class MonCommandCompletion : public Context
       Py_DECREF(python_completion);
       python_completion = nullptr;
     }
-    py_modules->notify_all("command", tag);
+    // do notify async since notify_all() takes py_modules' lock and we hold
+    // the MonClient or Objecter lock, and we don't want to deadlock
+    py_modules->notify_all_async("command", tag);
```

Comment:
Alternatively, instead of introducing `notify_all_async`, it seems you could just pass `MonCommandCompletion` context to the `finisher` (where  we call `command_c->finish(command_r)`).

Interestingly, why we call `finish` instead of `complete` there? It seems the `MonCommandCompletion` destructor is never called. Aren't we leaking?

Reply:
The parent Context complete() method is defined as ``{ finish(r); delete this; }``

Reply:
Yes, but we don't call `complete`, we call `finish` here [1]. And my question: is it intentional, and if it is then why? Because I don't see where `delete` is called in this case.

[1] https://github.com/ceph/ceph/blob/5f70363c68fe64f71afd4243777730fd9f47f101/src/mgr/BaseMgrModule.cc#L159